server/certs.go

```diff
@@ -8,52 +8,124 @@ import (
 	"bytes"
 	"crypto/sha256"
 	"crypto/tls"
+	"crypto/x509"
+	"fmt"
 	"io"
 	"os"
 	"time"
 
+	"github.com/fsnotify/fsnotify"
+
+	"github.com/open-policy-agent/opa/internal/pathwatcher"
 	"github.com/open-policy-agent/opa/logging"
 )
 
 func (s *Server) getCertificate(h *tls.ClientHelloInfo) (*tls.Certificate, error) {
-	s.certMtx.RLock()
-	defer s.certMtx.RUnlock()
+	s.tlsConfigMtx.RLock()
+	defer s.tlsConfigMtx.RUnlock()
 	return s.cert, nil
 }
 
-func (s *Server) certLoop(logger logging.Logger) Loop {
-	return func() error {
-		for range time.NewTicker(s.certRefresh).C {
-			certHash, err := hash(s.certFile)
+func (s *Server) reloadTLSConfig(logger logging.Logger) error {
+	s.tlsConfigMtx.Lock()
+	defer s.tlsConfigMtx.Unlock()
+
+	// if the server has a cert configured, then we need to check the cert and key for changes.
+	if s.certFile != "" {
+		certHash, err := hash(s.certFile)
+		if err != nil {
+			return fmt.Errorf("failed to check server certificate: %w", err)
+		}
+
+		certKeyHash, err := hash(s.certKeyFile)
+		if err != nil {
+			return fmt.Errorf("failed to check server key: %w", err)
+		}
+
+		different := !bytes.Equal(s.certFileHash, certHash) ||
+			!bytes.Equal(s.certKeyFileHash, certKeyHash)
+
+		if different { // load and store
+			newCert, err := tls.LoadX509KeyPair(s.certFile, s.certKeyFile)
 			if err != nil {
-				logger.Info("Failed to refresh server certificate: %s.", err.Error())
-				continue
+				return fmt.Errorf("failed to refresh server certificate: %w", err)
 			}
-			certKeyHash, err := hash(s.certKeyFile)
+			s.cert = &newCert
+			s.certFileHash = certHash
+			s.certKeyFileHash = certKeyHash
+			logger.Debug("Refreshed server certificate.")
+		}
+	}
+
+	// do not attempt to reload the ca cert pool if it has not been configured.
+	if s.certPoolFile != "" {
```

Comment:
If updating the server cert above fails, we'll never get to here.
A downside is that this might cause situations where the user can't update their client trust while having ongoing issues with the server cert. This could be confusing, as it looks like we'll fallback on an older, working server cert if there are issues.